### ISSUE TITLE
Could we start actively maintaining a .Net Core 3 build?

### ISSUE THREAD

**Issue by zur003** [APSIMInitiative/ApsimX#5344]
It's currently possible to compile all components other than the ApsimNG.exe user interface with .Net Core 3. @her123 has been working on using such builds in Docker containers for the World Modelers work, as they are more compact and much speedier than using Mono on Linux-based systems. Would it be possible to add this into the main branch in some way? I believe it is possible to have both Frameworks and Core projects in the same solution, though there may be some complications in doing so. It would be nice to ensure that a .Net Core version is always current, readily available, and able to run most or all of the test suite.

**Comment by hol430** [comment]
Good idea - maintaining a separate branch is going to be painful in the longer term. I have a netcore3 branch on my repo in which I have a somewhat-working gui on linux. The branch started out as your netcore3 branch but probably isn't quite up-to-date at the moment.

What would be the best way to merge this with the master branch? Do we just maintain a separate .csproj file for each solution?

**Comment by her123** [comment]
It would be nice to have the same code shared by two different solutions/projects. A dotnetcore solution may be able to live alongside the main solution in the main branch.

**Comment by hol430** [comment]
Thinking about it some more, I'm not sure if it will be possible to use exactly the same code for both solutions. The Gtk#3 api is a bit different to the gtk#2 one in some places, and I've had to make some changes which won't compile under gtk#2. No idea if you guys have had to make similarly breaking changes in the other projects to get them working in .net core.

**Comment by her123** [comment]
Is there too much to have in conditional sections?

**Comment by zur003** [comment]
Yes, what @her123 said is what I had in mind. The differences between Framework and .Net Core are very minor, if GUI code is excluded. It's just the code in Models\Core\ScriptCompiler that has any real differences.

**Comment by zur003** [comment]
More specifically, I was thinking we could have .Net core 3 builds of APSIM.Shared, APSIMRunner and Models, but not include any of the GUI stuff just yet. Not sure about UnitTests and Updater (Updater really only makes sense when driven by the GUI).

**Comment by hol430** [comment]
Ah, ok. Maybe we create two csproj files for each project except ApsimNG, then create a copy of ApsimNG which is based on .net core. That way we can still use the same code for framework/core for everything except ApsimNG?

**Comment by hol430** [comment]
Or yeah, leave the ApsimNG changes on a branch. Maybe that's the way to go.

**Comment by her123** [comment]
Sounds like a good way to go

**Comment by zur003** [comment]
One trick might be keeping the obj folder used for intermediate files separate for framework and core builds. It looks like this can be done by manual configuration of the .csproj files, but can't be done from the VS Gui.

**Comment by hol430** [comment]
Yes - good idea, manually purging those files each time I switch builds is going to get old fast.

Also, @zur003 do you recall why commit 078ade15c95341b8696da6326b93f7eb7371ae5b was necessary? It seems to break the running of .apsimx files for me when the .met file path contains %root%.

**Comment by zur003** [comment]
@hol430, commit 078ade1 was a hack because in the configuration I was using, the .net core binaries were going into bin\netcoreapp3.0 rather than bin. The root macro operates relative to the directory the Models.exe binary is in, so it needed to go up an additional level. Probably wasn't the best way to do it.
However, that raises the additional question of where the netcore binaries ought to go. If we build two different versions of Models.exe, they can't live in the same folder.

**Comment by hol430** [comment]
Ah ok, that makes sense. I've modified my netcore3 branch to allow for more convenient switching between core and framework builds. I've got 2 solution files, and 2 of each .csproj file, except for ApsimNG, of which I have 2 directories, each with 1 .csproj file. Switching between framework/core builds is achieved by just building the appropriate .sln file. A bit clunky maybe, but it seems to work (no manual deletion of obj/bin directories is necessary when switching back and forth). Getting the dependencies in place is a bit trickier, as I'm not using nuget for oxyplot or the sourceview components (yet), both of which need to be manually compiled.

**Comment by hol430** [comment]
I guess a single project file would be ideal, as long as it doesn't make building the main (framework) solution any harder for people who aren't programmers. Ideally, on windows, you could just select the appropriate framework/configuration from the drop-down in visual studio and hit build, no extra steps required.

**Comment by zur003** [comment]
@hol430 - You can now pull from my netcore3 branch to get project files for APSIM.Shared, Models, and APSIMRunner that build both Framework 4.7.2 and Core 3.0 targets. It took some rather obscure hand modifications to the .csproj files, but they are now configured to (1) keep the intermediate files from colliding with one another and (2) send the Framework binararies to the solution Bin folder and the Core binaries to a solution NetCoreBin folder. There's an option in the IDE that lets you select which framework is used when you're debugging.

Give it a try, and see whether this is what we should do in the master branch. I think it's working pretty well, although some care is needed when adding new nuget assemblies. It should be OK if the new assemblies support both frameworks, but can require manual editing of the .csproj file if they don't (this was necessary for the MigraDoc stuff). Also note that the Frameworks target is 4.7.2, rather than our current 4.6. You can use this approach with 4.6, but it ends up pulling dozens of System assemblies into the Bin folder. Can we support 4.7.2, or is it too "modern"?

**Comment by hol430** [comment]
Thanks @zur003, that's working for me. I don't really see any problem updating to .net 4.7.2, as long as the installer knows that we now require that particular version of .net.

I've just tried manually adding ApsimNG back into the .sln file, but it's not being built along with the other projects - any idea why that might be the case?

**Comment by hol430** [comment]
Sorry - ignore the last question. I forgot to add a build configuration to the solution file for ApsimNG. Turns out that reading the documentation can occasionally be a good idea.

What do you think of changing the output path (aka bin directory) for each project to `Bin/$(TargetFramework)` (so we would have `ApsimX/Bin/net472/Models.exe`). This seems to be a fairly common practice, and seems more intuitive to me than having a NetCoreBin directory, although it would break any existing user scripts which assume that Models.exe exists directly under Bin.

**Comment by zur003** [comment]
@hol430, That's actually the default way of doing things. I tinkered to make it behave differently, because I didn't want to have to deal with modifying the installers as well <g>. To get the default behaviour again, remove the line in each .csproj file with the  AppendTargetFrameworkToOutputPath element. You can also get rid of the line with IntermediateOutputPath. The line setting OutputPath to ..\Bin\ can then be moved into a PropertyGroup that does not have the net472 condition.

Have you been able to build ApsimNG? I'm still having problems with that.

**Comment by hol430** [comment]
Yeah, I'm not really sure what's best. I guess we should leave as is for now so that other people's scripts and the installers continue to work. As for ApsimNG, are you talking about the framework version -and what problems are you running into specifically? As far as I recall I just had to retarget it to .net 4.7.2, and add it back into the solution file [like so](https://github.com/APSIMInitiative/ApsimX/files/4843151/patch.txt).

**Comment by zur003** [comment]
@hol430 - my problem was that I was trying to redo the .csproj file in the "new" format, and didn't quite get it right. That's a move we'll probably want to make eventually, but it can wait a while longer. 

Do you think the multiple targeting is working well enough to push it into the master branch? The main problems I can see are: (1) this can make some project options less settable from the GUI; (2) adding new nuget packages could get messy if the package doesn't support both frameworks; (3) shifting to 4.7,2 may mean re-jigging installers a bit, and we'd need to test whether everything works properly under Mono on Linux and Mac.

**Comment by hol430** [comment]
Ah ok, yeah I haven't really tried converting ApsimNG to thew new sdk-style format. Agreed that it would be nice eventually.

I think it would be nice to get this into the master branch. That way we don't need to go out of our way to keep the branch up-to-date (except for .net core apsimng). I think problem 2 would be the biggest problem in the mid-term, but we don't really add new nuget packages all that often, so hopefully we can work around it on a case-by-case basis. I'm happy to look into modifying the installers and I'll also need to see about installing the appropriate .net core sdk on the build machines.

**Comment by hol430** [comment]
@zur003 @hol353 after just finishing a merge on my netcore3 branch, I'm wondering if having a separate set of directories/files for the .net core/gtk3 ApsimNG is really the best option, because any changes to ApsimNG need to be manually ported across to the netcore version of ApsimNG. git can do some of the legwork, but creating/fixing the patch files is a bit of a hassle. I'm wondering if it would be better to have both framework and netcore buidls of ApsimNG using the same files/code, with a bunch of `#if NETFRAMEWORK` directives scattered throughout the code. It will be ugly, but really there aren't many people who actually work on ApsimNG code and it might save us from having to fix bugs again down the line. Any thoughts?

**Comment by her123** [comment]
It's generally easier to maintain some conditional code sections than separate (duplicate) files. Use the least redundancy possible.

**Comment by hol353** [comment]
How much conditional code would there be in ApsimNG?

**Comment by hol430** [comment]
Quite a lot, I have over 1k compiler warnings on my .net core ApsimNG project, most of which are due to changes between gtk2/3 and will require some conditional code. For example, widget.Destroy() is deprecated in favour of widget.Dispose(), so we would need a conditional directive around each call to widget.Destroy() inside our code.

On the other hand, maintaining two completely separate sets of files is going to be a nightmare.

**Comment by hol353** [comment]
Ok go for it.

**Comment by zur003** [comment]
@hol430 - Why not write a static DestroyWidget function that gets used everywhere and call Destroy or Dispose appropriately in that function? That could reduce the number of conditionals considerably. As a general rule, if the "same" conditional logic needs to be used more than once, it would probably be better to consolidate into a single function somewhere, if possible.

**Comment by hol430** [comment]
True! After having another look it appears that a lot of the warnings are also due to missing xml documentation comments which won't require platform-specific changes. So it might not be too bad at all

**Comment by hol430** [comment]
@hol353, @zur003 - it appears that moving ApsimNG to .net 4.7.2 is causing problems with MarkdownDeep, our markdown processor, which targets .net 3.5. Actually, I'm not sure how this was working in our .net 4.6.1 builds on master branch either, but the MarkdownDeep project seems to be abandoned anyway, so perhaps it's time to change to another markdown processor such as [MarkDig](https://github.com/lunet-io/markdig)?

**Comment by zur003** [comment]
@hol430 - I'm happy if you switch. My testbed version of ApsimNG for Gtk#3 has been using MarkdownSharp. It may be worth shopping around a little, since these libraries differ a bit in the markdown features they support.

**Comment by hol430** [comment]
Yeah there seems to be quite a few different markdown libraries for .net. The main difference that I've noticed between them is that the older ones don't support [CommonMark](https://commonmark.org/) which is the newer, more precise/strict markdown specification which is used by github, stack overflow and others.

As far as features go, it would be handy if we could somehow specify a resource name as an image URI. There might be some others that I've forgotten about as well.

**Comment by hol430** [comment]
@hol353, @zur003 what do you think is the next step on this? We could probably merge this into the master branch at some point, the main obstacle that I see is that it would require all developers to install the .net core sdk. The visual studio experience is that it builds both netcore and framework versions, and when you hit run in visual studio it will run the .net framework version, so most people shouldn't be affected. The .net framework build seems pretty stable to me - it is essentially doing the same thing as the current master branch except on .net 4.7.2 instead of .net 4.6(.1?).

**Comment by hol353** [comment]
If we merge this into master will VS prompt developers to install .net core?

**Comment by hol430** [comment]
No, most likely builds will fail with a cryptic message about a missing .net core targeting pack.

**Comment by hol353** [comment]
That's unfortunate. So the problem will be that the developers will start sending emails saying they can't compile. 

Can we get a list of email addresses from GitHub of all the users who have forked the repo? If so then we can send an email out.

**Comment by sarahcleary** [comment]
Anyone that logs a question or I notice is an active participant on the APSIM repository, I've been inviting to become a member. Not sure if that covers all of the people you need to get to - and also, not sure if you can tag the entire membership group - and hence, they all receive a message.  Could get them to join particular team, and thus tag the team (and then they would receive an email notification via GitHub)?

**Comment by zur003** [comment]
I haven't tested this on LInux or OS X. Will builds still work on those platforms?

**Comment by hol430** [comment]
I've been doing most of my testing on Linux, and it works well for me, better than the mono/framework builds in fact (apart from the parts of the GUI which haven't been implemented yet of course). I've been using vs code for debugging and it is very easy to set up, and it also runs on windows and mac. No idea about how if monodevelop will build the sdk-style projects.

As for getting people compiling, I believe you can sticky/pin a github issue so that it appears separate to the other issues for better visibility. E.g. this repo has to stickied issues [this repo](https://github.com/dotnet/runtime/issues). We could create such an issue pointing people to the updated compiling instructions, and telling them they need to install the .net core sdk, then just ping every team in the APSIMInitiative organisation.

**Comment by hol353** [comment]
> Anyone that logs a question or I notice is an active participant on the APSIM repository, I've been inviting to become a member. Not sure if that covers all of the people you need to get to - and also, not sure if you can tag the entire membership group - and hence, they all receive a message. Could get them to join particular team, and thus tag the team (and then they would receive an email notification via GitHub)?

Yes I agree. I've renamed the 'Fixers' team to 'Developers'. It has most people that we need to contact. When the time comes we can tag that team to an issue (this issue?) with instructions on what they need to do to install .Net Core. No need to tag all teams.

**Comment by hol430** [comment]
I will need to update the compiling instructions on the netlify site, so we could just point them over that way. And probably mention that they need to install .net core sdk. Making this move will also mean dropping support for vs 2017.

**Comment by hol430** [comment]
@hol353 it appears that using soap services is considered deprecated these days, and your recent work on the soil downloader is not going to work in .net core builds. For instance, the System.Web.Services namespace is unavailable. How hard will it be to refactor this to find an alternate (e.g. rest) way of doing things?

**Comment by hol353** [comment]
Hmm we would have to create an APSoil rest service and host it on apsimdev. Not hard but not a priority for me atm.

**Comment by hol430** [comment]
Yeah, makes sense. For now I have just disabled this feature in the .net core builds.

**Comment by zur003** [comment]
Guess I really am getting old. It doesn't seem that long along that SOAP was planning to take over the world. In any case, @hol353, if this is revised, please remember that the SoilMapp app is using the SOAP service.

**Comment by hol353** [comment]
I won't get rid of the old APSoil service. I'll create a new REST one.

**Comment by hol430** [comment]
@hol353 - after #5998, the only place that the htmlview is still used is in the upgrade UI, where it's used to display the license agreement. I suspect we can probably use a markdownview here instead, but the problem is that the license agreement document is in .docx format. I've managed to somewhat successfully convert the .docx file to markdown, but I'll need to make some changes to the .docx file to improve the markdown output (e.g. some of the regular text in this file is in a heading style/format which looks fine in a word processor but not in markdown). Alternatively, we could just display the lincense file stored in this repo, which is already conveniently in markdown format. Let me know which you would prefer.

**Comment by hol353** [comment]
Happy for you to use the markdown version if you remove the .docx file. We'll have to remember though that when @sarahcleary updates the APSIM license we will need to convert the .docx to markdown and store that in the repo.

**Comment by hol353** [comment]
Some bugs with new .net core GTK3 GUI:

- When you right click on a manager you don't get a menu with 'Test compile', 'Reformat' etc.
- The GUI is really, really slow copying and pasting nodes in the tree (or adding a new model). It throws lots of Exceptions if you're running in VS. The  exceptions are recoverable and the GUI does actually do what you want, eventually.

**Comment by hol430** [comment]
- Progress bar doesn't expand to fill available (vertical) space

**Comment by hol353** [comment]
I can't generate auto-doc. e.g. Open soybean validation, right click on Simulations and choose Create Documentation. Get this:
![image](https://user-images.githubusercontent.com/1943827/107174507-1fc8a500-6a16-11eb-9ef6-b642c1ecf3c6.png)
Graph.Width and Graph.Height are zero.

**Comment by hol353** [comment]
Graph series view is missing the labels:

![image](https://user-images.githubusercontent.com/1943827/107428312-2c601100-6b6e-11eb-8ae3-108d72e90a05.png)

**Comment by hol353** [comment]
Graph folders are mangled - see chickpea.apsimx:

![image](https://user-images.githubusercontent.com/1943827/107428545-86f96d00-6b6e-11eb-9d7f-cf9b548d8944.png)

**Comment by hol353** [comment]
When generating a pdf document (autodoc), and exception is thrown. The chickpea.pdf file that is being opened exists and is a valid pdf file.
![image](https://user-images.githubusercontent.com/1943827/107429114-4c440480-6b6f-11eb-9405-046f3baeee97.png)

**Comment by hol353** [comment]
The generated PDF document isn't rendering subscripts properly. It is putting ~ characters in (e.g. XF~1~):

![image](https://user-images.githubusercontent.com/1943827/107429713-f1f77380-6b6f-11eb-9bfa-bb5f0a91d254.png)

**Comment by hol353** [comment]
The generated PDF has a problem scaling one of the images - see below. The other images seem ok:
![image](https://user-images.githubusercontent.com/1943827/107430505-f708f280-6b70-11eb-92a3-9f516aeedbeb.png)

**Comment by hol353** [comment]
When a validation graph gets written to the pdf document the regression line, 1:1 lines and the stats aren't visible. I'm using a dark mode in the GUI and writing to a white background PDF.
![image](https://user-images.githubusercontent.com/1943827/107431325-0472ac80-6b72-11eb-87bf-9ac457a4e287.png)

**Comment by hol353** [comment]
The map view in the GUI is not showing the lat/long properties. You can drag the splitter down to see them but shouldn't have to do this:
![image](https://user-images.githubusercontent.com/1943827/107431582-629f8f80-6b72-11eb-9587-e0d94c64ec67.png)

**Comment by hol353** [comment]
While editing a cultivar in the GUI, the highlighted line is very difficult to read:

![image](https://user-images.githubusercontent.com/1943827/107435040-5a961e80-6b77-11eb-8770-2b50e73b9a52.png)

**Comment by hol353** [comment]
When running the chickpea validation I get an error in ExcelInput.cs. Only seems to happen with .net core version.

![image](https://user-images.githubusercontent.com/1943827/107440976-3ee34600-6b80-11eb-914d-cd756a06e675.png)

**Comment by hol353** [comment]
Intellisense doesn't seem to work. I was looking at wheat example. In Report, I typed '[Wheat].' and got the exception below.

![image](https://user-images.githubusercontent.com/1943827/107448526-8c19e480-6b8d-11eb-8af9-64521925e334.png)

**Comment by hol353** [comment]
The icons in the toolbar and in the tree view look small to me when compared to the 'Home' screen
![image](https://user-images.githubusercontent.com/1943827/108911153-a0241280-7672-11eb-9622-305e318398f7.png)

**Comment by hol353** [comment]
I can no longer do a search in manager scripts i.e. Ctrl F doesn't work.

**Comment by hol430** [comment]
Closing in favour of #6139.
